src/profile-logic/process-profile.js

```diff
@@ -1059,53 +1066,193 @@ export function processProfile(
+ */
+export function sanitizePII(
+  profile: Profile,
+  PIIToBeRemoved: RemoveProfileInformation
+): Profile {
   let urlCounter = 0;
+  const removedThreadIndexes = [];
   const newProfile = Object.assign({}, profile, {
-    meta: { ...profile.meta, networkURLsRemoved: !includeNetworkUrls },
+    meta: {
+      ...profile.meta,
+      // FIXME: We won't be needing this after PII frontend changes.
+      networkURLsRemoved: PIIToBeRemoved.shouldRemoveNetworkUrls,
+      extensions: PIIToBeRemoved.shouldRemoveExtensions
+        ? getEmptyExtensions()
+        : profile.meta.extensions,
+    },
     pages:
-      includeNetworkUrls === false && profile.pages
+      (PIIToBeRemoved.shouldRemoveNetworkUrls ||
+        PIIToBeRemoved.shouldRemoveAllUrls) &&
+      profile.pages
         ? profile.pages.map(page =>
             Object.assign({}, page, {
               url: 'Page #' + urlCounter++,
             })
           )
         : profile.pages,
-    threads: profile.threads.map(thread => {
-      const stringArray = thread.stringTable.serializeToArray();
-      const newThread = Object.assign({}, thread);
-      delete newThread.stringTable;
-      if (includeNetworkUrls === false) {
-        for (let i = 0; i < newThread.markers.length; i++) {
-          const currentMarker = newThread.markers.data[i];
-          if (
-            currentMarker &&
-            currentMarker.type &&
-            currentMarker.type === 'Network'
-          ) {
-            // Remove the URI fields from marker payload.
-            currentMarker.URI = '';
-            currentMarker.RedirectURI = '';
-            // Strip the URL from the marker name
-            const stringIndex = newThread.markers.name[i];
-            stringArray[stringIndex] = stringArray[stringIndex].replace(
-              /:.*/,
-              ''
-            );
-          }
-        }
+    threads: profile.threads.reduce((acc, thread, threadIndex) => {
+      let newThread: any = Object.assign({}, thread);
+
+      newThread = sanitizeThreadPII(newThread, threadIndex, PIIToBeRemoved);
+
+      if (newThread === null) {
+        removedThreadIndexes.push(threadIndex);
+        // Filtering out the current thread if it's null.
+        return acc;
       }
-      newThread.stringArray = stringArray;
-      return newThread;
-    }),
+
+      // Adding the thread to the `threads` list.
+      return acc.concat(newThread);
+    }, []),
+    // Remove counters which belong to the removed counters.
+    // Also adjust other counters to point to the right thread.
+    counters: profile.counters
+      ? profile.counters.reduce((acc, counter) => {
+          if (removedThreadIndexes.includes(counter.mainThreadIndex)) {
+            // Filtering out the current counter.
+            return acc;
+          }
+
+          let reduceBy = 0;
+          for (const threadIndex of removedThreadIndexes) {
+            if (counter.mainThreadIndex > threadIndex) {
+              reduceBy++;
```

Comment:
I'm not sure this is correct. I think we need to do a `Map<OldIndex, NewIndex | null>`.